bindings: counter: extend binding with channel specifier

javlands opened the pull request:
**Suggestion**: make it possible to define the channel in the device tree when using a hardware counter.

**Proposal**: add a _counter.yaml_ file that specifies a #counter-cells property. This can than be included and used to add the channel to the specifiers.

**Fixes**: drivers that make use of hardware timers can now get the channel from the device tree, making them more portable.

As this change affects a large portion of the tree, I only adapted thenative sim and st,stm32-counter compatible for now to keep this PR organized. 

**Usecase**: When writing bitbang drivers, hardware counters can be used to toggle the gpio's (instead of a blocking wait like the i2c_bitbang). The counter channel can then be specified in the device tree.

I'd love to hear your feedback on this change.

erwango reviewed (dismissed):
Please split generic binding change and vendors adaptation in specific commits and more importantly please provide a usage example.

javlands committed:
bindings: counter: extensibility with channel specifier

It was not yet possible to specify in the device tree which channel a
counter instance should use. This made drivers using counters less
portable. This is now fixed by adding a new counter.yaml file that
adds the #counter-cells property. This file can be included in the
counter binding.

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

javlands committed:
dts: arm: st: extended counter dts binding for st boards

Add the channel specifier to the counter dts binding for st boards
and adapt all affected boards

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

javlands committed:
boards: native: native_sim: extended counter dts binding for native sim

Add the channel specifier to the counter dts binding for native
sim and adapt affected board

Signed-off-by: Jaro Van Landschoot <jaro.vanlandschoot@basalte.be>

erwango commented:
> Usecase: When writing bitbang drivers, hardware counters can be used to toggle the gpio's (instead of a blocking wait like the i2c_bitbang). The counter channel can then be specified in the device tree.

Thanks for the description update which I had missed. Though, by `please provide a usage example.`, I meant a code sample.  Would you have one ?
This being said I fail to see if this would break existing users or if this could only be used "on top" of existing usages. I guess this is one thing to clarify

Also, it would be great to have @nordic-krch feedback.

Removing "STM32" tag to open the discussion.

javlands commented:
> Thanks for the description update which I had missed. Though, by `please provide a usage example.`, I meant a code sample. Would you have one ? 

I don't have a code sample, as these changes mostly affect drivers that use hardware timers. I can however sketch you a usecase:
Imagine you want to design a full-duplex bitbanged transeiver using hardware timers. Your device tree node could then look something like the following:

```
bitbang_bus: bitbang-bus {
	status = "okay";
	compatible = "my,compatible"
	#address-cells = <1>;
	#size-cells = <0>;
	baudrate = <15000>;

	rx-gpios = <&gpio0 0 0>;
	tx-gpios = <&gpio0 1 0>;
	counters = <&counter0 0>, <&counter0 1>;	
}
```

The driver will consist of 3 main parts:
 - A tx counter callback, scheduled periodically to toggle the tx gpio for transmission
 - A rx gpio interrupt, used to detect the start of incoming data (neglecting clockdrift for simplicity). This interrupt starts the rx counter so it is triggered at the desired sample point
 - A rx counter callback, sampling each time it is triggered, and rescheduling itself for the next data symbol.
 
The main advantage of adding the channels as counter-cells, is that when a hardware counter supports multiple channels, these can be specified in the device tree, instead of in the driver itself, increasing portability.

A real life scenario would be a Dali bitbanged bus, like described here (https://github.com/zephyrproject-rtos/zephyr/issues/71544). Using busy waits will block the cpu for too long of a time due to the low baudrate of dali (1200).


> This being said I fail to see if this would break existing users or if this could only be used "on top" of existing usages. I guess this is one thing to clarify

It does not break anything in the Zephyr source code, as there isn't a driver that uses a hardware counter as a device tree node property.

erwango commented:
Thanks @javlands for the context, this really helps. This sounds quite reasonable to me.
However, to get more feedbacks, I'd suggest to open an issue (with basically the same description of the overall approach). This will provide more visibility to your proposal and speed things up (adding tag RFC will make it appear here: https://github.com/orgs/zephyrproject-rtos/projects/24, which doesn't work with PR)

nashif commented:
what is the status of this PR? we keep removing the stale bot, yet, nothing happens.

pdgendt commented:
> what is the status of this PR? we keep removing the stale bot, yet, nothing happens.

It's still on our list, but with a lower priority. Since this already has merge conflicts we'll open a new PR when resuming.
